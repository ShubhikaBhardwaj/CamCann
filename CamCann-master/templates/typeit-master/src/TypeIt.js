import Instance from "./Instance";
import allHaveStatus from "./helpers/allHaveStatus";
import { generateHash } from "./utilities";
import toArrayOfNodes from "./helpers/toArrayOfNodes";

export default class TypeIt {
  constructor(element, options) {
    this.instances = toArrayOfNodes(element).map(el => {
      return new Instance({
        typeIt: this,
        element: el,
        id: generateHash(),
        options,
        queue: []
      });
    });
  }

  each(func) {
    this.instances.forEach(instance => {
      func.call(this, instance);
    });
  }

  /**
   * Push a specific action directly into the queue of each instance.
   * If an instance has already completed, trigger the queeu again.
   *
   * @param {string} function
   * @param {*} argument
   * @
   */
  queueUp(action, argument = null, numberOfTimesToCopy = 1) {
    this.each(instance => {
      let isIndependentFunction = typeof action !== "string";

      /**
       * If action is a string, bind it to instance.
       * Otherwise, leave it on its own.
       */
      let toFire = isIndependentFunction ? action : instance[action];

      let toPassAsArguments = isIndependentFunction ? this : argument;

      for (let i = 0; i < numberOfTimesToCopy; i++) {
        instance.queue.add([toFire, toPassAsArguments]);
      }
    });
  }

  is(status) {
    return allHaveStatus(this.instances, status, true);
  }

  freeze() {
    this.each(instance => {
      instance.status.frozen = true;
    });
  }

  unfreeze() {
    this.each(instance => {
      if (!instance.status.frozen) return;
      instance.status.frozen = false;
      instance.fire();
    });
  }

  /**
   * If used after typing has started, will append strings to the end of the existing queue. If used when typing is paused, will restart it.
   *
   * @param  {string} string The string to be typed.
   * @return {object} TypeIt instance
   */
  type(string = "") {
    this.each(instance => instance.queueString(string));
    return this;
  }

  /**
   * If null is passed, will delete whatever's currently in the element.
   *
   * @param  { number } numCharacters Number of characters to delete.
   * @return { TypeIt }
   */
  delete(numberOfCharactersToDelete = null) {
    this.queueUp(
      "delete",
      numberOfCharactersToDelete === null, //-- Maybe delete all.
      numberOfCharactersToDelete === null ? 1 : numberOfCharactersToDelete
    );

    return this;
  }

  pause(ms = null) {
    this.queueUp("pause", ms);
    return this;
  }

  break() {
    this.queueUp("type", document.createElement("BR"));
    return this;
  }

  options(options) {
    this.queueUp("setOptions", options);
    return this;
  }

  exec(func) {
    this.queueUp(func);
    return this;
  }

  /**
   * Destroy the instance, mark as such, and clean up. If specified,
   * remove the cursor on destruction.
   *
   * @param {boolean} removeCursor
   */
  destroy(removeCursor = true) {
    this.each(instance => {
      // Destroy each timeout, for good housekeeping.
      instance.timeouts.forEach(timeout => {
        clearTimeout(timeout);
      });

      instance.timeouts = [];

      // Clean up cursor node, if specified.
      if (removeCursor) {
        let cursorNode = instance.isInput
          ? null
          : instance.$e.querySelector(".ti-cursor");

        if (cursorNode) {
          instance.$e.removeChild(cursorNode);
        }
      }

      instance.status.destroyed = true;
    });
  }

  empty() {
    this.queueUp("empty");
    return this;
  }

  /**
   * Reset each instance like it's brand new.
   */
  reset() {
    this.destroy();

    this.instances = this.instances.map(instance => {
      return instance.reset();
    });

    return this;
  }

  /**
   * Initialize each instance.
   */
  go() {
    this.each(instance => {
      instance.init();
    });

    return this;
  }
}
