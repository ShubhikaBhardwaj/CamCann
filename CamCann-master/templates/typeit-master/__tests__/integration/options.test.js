import TypeIt from "../../src/TypeIt";

test("Cursor should function by default.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {
    strings: ["This should have a default cursor."]
  }).go();

  let cursorCharacter = document
    .getElementById("element")
    .querySelector(".ti-cursor").innerHTML;

  expect(cursorCharacter).toBe("|");
});

test("Changes cursor character correctly.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  const instance = new TypeIt("#element", {
    strings: ["This should have a custom cursor."],
    cursorChar: "$"
  }).go();

  let cursorCharacter = document
    .getElementById("element")
    .querySelector(".ti-cursor").innerHTML;

  expect(cursorCharacter).toBe("$");
});

test("Turning off cursor should work.", () => {
  document.body.innerHTML = `<div>'
      <span id="element"></span>
    </div>`;

  new TypeIt("#element", {
    strings: ["This should have no cursor."],
    cursor: false
  });

  let visibilityStyle = document.getElementById("element").style.visibility;

  expect(visibilityStyle).toBe("");
});

describe("startDelete option.", () => {
  test("Makes hard-coded string the first string to be typed.", () => {
    document.body.innerHTML = `<div>'
      <span id="element">This should be typed first.</span>
    </div>`;

    const instance = new TypeIt("#element", {});

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();

    expect(instance.instances[0].opts.strings).toEqual([
      "This should be typed first."
    ]);
  });

  test("Sets both hard-coded and option-defined strings in correct order.", () => {
    document.body.innerHTML = `<div>'
      <span id="element">This should be typed first.</span>
    </div>`;

    const instance = new TypeIt("#element", {
      strings: "This is another string."
    });

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();

    expect(instance.instances[0].opts.strings).toEqual([
      "This should be typed first.",
      "This is another string."
    ]);
  });

  test("Correctly adds strings also defined by type() companion method.", () => {
    document.body.innerHTML = `<div>'
      <span id="element">This should be typed first.</span>
    </div>`;

    const instance = new TypeIt("#element", {
      strings: "This is another string."
    }).type("And finally, a third.");

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();

    expect(instance.instances[0].opts.strings).toEqual([
      "This should be typed first.",
      "This is another string."
    ]);
  });
});

describe("html option", () => {
  test("It should queue strings as HTML by default.", () => {
    document.body.innerHTML = `
      <div>
        <span id="element"></span>
      </div>
    `;

    const instance = new TypeIt("#element", {
      strings: "This is a <strong>BOLD</strong> string."
    });

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });

  test("It should queue strings as HTML when explicitly set.", () => {
    document.body.innerHTML = `
      <div>
        <span id="element"></span>
      </div>
    `;

    const instance = new TypeIt("#element", {
      strings: "This is a <strong>another BOLD</strong> string.",
      html: true
    });

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });

  test("It should leave strings be when option is disabled.", () => {
    document.body.innerHTML = `
      <div>
        <span id="element"></span>
      </div>
    `;

    const instance = new TypeIt("#element", {
      strings: "This is a <strong>another BOLD</strong> string.",
      html: false
    });

    expect(instance.instances[0].queue.waiting).toMatchSnapshot();
  });
});
