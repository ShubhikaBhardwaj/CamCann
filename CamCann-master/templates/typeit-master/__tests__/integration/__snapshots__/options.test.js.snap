// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`html option It should leave strings be when option is disabled. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "<",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ">",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "B",
  ],
  Array [
    [Function],
    "O",
  ],
  Array [
    [Function],
    "L",
  ],
  Array [
    [Function],
    "D",
  ],
  Array [
    [Function],
    "<",
  ],
  Array [
    [Function],
    "/",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ">",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
]
`;

exports[`html option It should queue strings as HTML by default. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "B",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "O",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "L",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "D",
    },
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
]
`;

exports[`html option It should queue strings as HTML when explicitly set. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "a",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "n",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "o",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "t",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "h",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "e",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "r",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": " ",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "B",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "O",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "L",
    },
  ],
  Array [
    [Function],
    Object {
      "ancestorTree": Array [
        "STRONG",
      ],
      "attributes": Array [],
      "content": "D",
    },
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
]
`;

exports[`startDelete option. Correctly adds strings also defined by type() companion method. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "u",
  ],
  Array [
    [Function],
    "l",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "b",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "y",
  ],
  Array [
    [Function],
    "p",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "f",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
  Array [
    [Function],
    375,
  ],
  Array [
    [Function],
    <br />,
  ],
  Array [
    [Function],
    375,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
  Array [
    [Function],
    "A",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "f",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    "l",
  ],
  Array [
    [Function],
    "l",
  ],
  Array [
    [Function],
    "y",
  ],
  Array [
    [Function],
    ",",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
]
`;

exports[`startDelete option. Makes hard-coded string the first string to be typed. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "u",
  ],
  Array [
    [Function],
    "l",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "b",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "y",
  ],
  Array [
    [Function],
    "p",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "f",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
]
`;

exports[`startDelete option. Sets both hard-coded and option-defined strings in correct order. 1`] = `
Array [
  Array [
    [Function],
    250,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "u",
  ],
  Array [
    [Function],
    "l",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "b",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "y",
  ],
  Array [
    [Function],
    "p",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    "d",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "f",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
  Array [
    [Function],
    375,
  ],
  Array [
    [Function],
    <br />,
  ],
  Array [
    [Function],
    375,
  ],
  Array [
    [Function],
    "T",
    Object {
      "isFirst": true,
    },
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "a",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "o",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "h",
  ],
  Array [
    [Function],
    "e",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    " ",
  ],
  Array [
    [Function],
    "s",
  ],
  Array [
    [Function],
    "t",
  ],
  Array [
    [Function],
    "r",
  ],
  Array [
    [Function],
    "i",
  ],
  Array [
    [Function],
    "n",
  ],
  Array [
    [Function],
    "g",
  ],
  Array [
    [Function],
    ".",
    Object {
      "isLast": true,
    },
  ],
]
`;
